### PR TITLE
147 optimize metrics for better compression

### PR DESCRIPTION
**Pull request by lquerel**
This PR applies the optimizations that were done on traces and logs to the metrics OTel entities. The following screenshots provide a comparison between `OTLP Dict` and `OTLP Arrow` for both `hipstershop_metrics.pb` and `mutlivariate_metrics.pb`.

In the case of `hipstershop_metrics.pb`, stream mode proves to be decidedly more effective when dealing with batches exceeding 500 metrics (probably equivalent around 250 entities). As anticipated, this advantage only becomes evident with larger batches in unary rpc mode, given the schema overhead requires substantial data to be amortized on the “OTLP Arrow” side.

Screenshot 2023-05-10 at 1 12 22 PM

For `multivariate_metrics.pb`, which aligns closely with F5's use of OTEL, the stream mode consistently outperforms for `OTLP Arrow`. This holds true in batch mode for batches exceeding 100 metrics.

Screenshot 2023-05-10 at 1 14 56 PM

The general Arrow data model used for the metrics is summarized with the following diagram.

OTLP Arrow Adapter - metrics

This mapping splits the hierarchical OTel model into 17 distinct Arrow records in order to apply different specific row order based on the specificities of each record type (e.g. row order for attributes is different of the row order for summaries). This new representation provides a better overall compression ratio.

In addition metrics are no longer represented with a union (i.e. union of sum, gauge, summary, histogram, and exp_histogram) as they are now split in individual records. A future PR will remove the remaining unions in order to simplify the bridge with the Parquet format.

Shared attributes are no longer processed individually, as the new way of organizing and sorting attributes automatically compresses common attributes between metrics. This saves time and resources. The same approaches will be applied to the logs and traces in a future PR.

### CODE REVIEW

File: pkg/arrow/from_record.go [hunk 1]
```diff
@@ -119,7 +285,43 @@ func StringFromRecord(record arrow.Record, fieldID int, row int) (string, error)
 	return StringFromArray(arr, row)
 }
 
-// SparseUnionFromRecord returns the sparse union value for a specific row and
+// BinaryFromRecord returns the []byte value for a specific row and column in
+// an Arrow record. If the value is null, it returns nil.
+func BinaryFromRecord(record arrow.Record, fieldID int, row int) ([]byte, error) {
+	if fieldID == -1 {
```
Comment (jmacd) [inline pkg/arrow/from_record.go:291]: I wonder if this `-1` should be a constant somewhere, to help readability (like `AbsentFieldID = -1`)?

File: pkg/otel/metrics/arrow/ehistogram_dp.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+		{Name: constants.ExpHistogramNegative, Type: EHistogramDataPointBucketsDT, Metadata: schema.Metadata(schema.Optional)},
+		{Name: constants.Exemplars, Type: arrow.ListOf(ExemplarDT), Metadata: schema.Metadata(schema.Optional)},
+		{Name: constants.Flags, Type: arrow.PrimitiveTypes.Uint32, Metadata: schema.Metadata(schema.Optional)},
+		{Name: constants.HistogramMin, Type: arrow.PrimitiveTypes.Float64, Metadata: schema.Metadata(schema.Optional)},
+		{Name: constants.HistogramMax, Type: arrow.PrimitiveTypes.Float64, Metadata: schema.Metadata(schema.Optional)},
+	}, nil)
+)
+
+type (
+	// EHistogramDataPointBuilder is a builder for exponential histogram data points.
```
Comment (jmacd) [inline pkg/otel/metrics/arrow/ehistogram_dp.go:63]: It almost looks like some of the code could be auto-generated, not that I'd want to wait for that. :grin:

**Review by jmacd — APPROVED** [review]
I'm starting to feel that I would have trouble maintaining this code without you, but the PR description was helpful and at a high-level I understand what's going on here. Looks good.

File: pkg/arrow/from_record.go (hunk 1 above)
Comment (lquerel) [reply pkg/arrow/from_record.go]: Good idea. Added this constant in https://github.com/f5/otel-arrow-adapter/pull/149/commits/842afdea45f6ff1874dc55b60bd34a133f2d6ae0

File: pkg/otel/metrics/arrow/ehistogram_dp.go (hunk 2 above)
Comment (lquerel) [reply pkg/otel/metrics/arrow/ehistogram_dp.go:74]: Fully agree. I'm missing the Rust macro system for this kind of repetitive code. If I have enough energy/time I'll try to use the Go code generation framework to simplify this code.

**Pull request merged**